File: backend/analyzer.py
import pandas as pd
import re
from collections import Counter
import emoji

# --- SABİTLER VE FİLTRELER ---

STOP_WORDS = set([
    "bir", "ve", "ile", "de", "da", "bu", "şu", "o", "ben", "sen", "biz", "siz", "onlar",
    "mı", "mi", "mu", "mü", "ama", "fakat", "lakin", "ancak", "için", "gibi", "kadar",
    "diye", "yok", "var", "ne", "nasıl", 'neden', 'niye', 'şey', 'yani', 'zaten',
    'tamam', 'peki', 'evet', 'hayır', 'olur', 'olmaz', 'iyi', 'güzel', 'kötü',
    'bunu', 'şunu', 'onu', 'bana', 'sana', 'ona', 'şimdi', 'sonra', 'önce', 'daha',
    'en', 'çok', 'az', 'biraz', 'bile', 'ise', 'tüm', 'her', 'hiç', 'kendi',
    'sadece', 'veya', 'ya', 'hem', 'böyle', 'öyle', 'zaman', 'olsun', 'olsa',
    'hani', 'işte', 'falan', 'filan', 'tmm', 'ok', 'oke', 'aynen', 'tabii', 'tabi',
    'değil', 'misin', 'musun', 'bence', 'sence', 'galiba', 'sanırım', 'bide',
    'artık', 'hala', 'diğer', 'mesaj', 'kanka', 'abi', 'abla', 'di', 'bi', 'be', 'lan'
])

SYSTEM_FILTERS = [
    "dahil edilmedi", "uçtan uca şifreli", "araması", "kişisini ekledi",
    "grubu oluşturdu", "gruptan ayrıldı", "tanımı değişti", "sildi",
    "silindi", "düzenlendi", "iletilen", "konumu paylaştı", "kartvizit"
]

BLACKLIST_BIGRAMS = [
    "dahil edilmedi", "çıkartma dahil", "mesaj düzenlendi",
    "mesaj bekleniyor", "mesaj silindi", "bu mesaj", "bu ileti",
    "uçtan uca", "uca şifreli", "araması cevapsız", "görüntülü arama",
    "bekleniyor alabilir", "zaman alabilir", "işlem biraz",
    "ses dahil", "video dahil", "görüntü dahil",
    "mesajı sildiniz", "sildiniz", "hadi demek"
]

class WhatsAppAnalyzer:
    def __init__(self, csv_path):
        self.df = pd.read_csv(csv_path)
        self._preprocess_data()

    def _preprocess_data(self):
        # Tarih ve Saat birleştirme
        self.df['Tarih_Formatli'] = pd.to_datetime(self.df['Tarih'], format='%d.%m.%Y', errors='coerce')
        self.df['Tam_Zaman'] = pd.to_datetime(
            self.df['Tarih'] + ' ' + self.df['Saat'],
            format='%d.%m.%Y %H:%M:%S',
            errors='coerce'
        )
        self.df = self.df.sort_values('Tam_Zaman')
        self.df['Saat_Sayisal'] = self.df['Saat'].str[:2].astype(int)

    def get_message_stats(self):
        """Mesaj sayıları."""
        msg_counts = self.df['Gonderen'].value_counts().reset_index()
        msg_counts.columns = ['Kişi', 'Mesaj Sayısı']
        return msg_counts.to_dict('records')

    def get_top_words(self, top_n=3):
        """En çok kullanılan kelimeler."""
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str)
            valid_words = []
            for msg in user_msgs:
                msg_lower = msg.lower()
                if any(sys_filter in msg_lower for sys_filter in SYSTEM_FILTERS): continue
                msg_clean = re.sub(r'[^\w\s]', '', msg_lower)
                words = msg_clean.split()
                for w in words:
                    if w not in STOP_WORDS and len(w) > 3:
                        valid_words.append(w)
            most_common = Counter(valid_words).most_common(top_n)
            result[user] = [{"word": word, "count": count} for word, count in most_common]
        return result

    def get_emoji_stats(self, top_n=5):
        """En çok kullanılan emojiler."""
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str)
            all_emojis = []
            for msg in user_msgs:
                extracted = [item['emoji'] for item in emoji.emoji_list(msg)]
                all_emojis.extend(extracted)
            if all_emojis:
                top_emojis = Counter(all_emojis).most_common(top_n)
                result[user] = [{"emoji": e, "count": c} for e, c in top_emojis]
            else:
                result[user] = []
        return result

    def get_hourly_activity(self):
        """Saat dilimi analizi."""
        bins = [0, 3, 6, 9, 12, 15, 18, 21, 24]
        labels = [
            "00:00 - 03:00 (Gece)", "03:00 - 06:00 (Sabaha Karşı)", 
            "06:00 - 09:00 (Sabah)", "09:00 - 12:00 (Kuşluk)", 
            "12:00 - 15:00 (Öğle)", "15:00 - 18:00 (İkindi)", 
            "18:00 - 21:00 (Akşam)", "21:00 - 00:00 (Gece Yarısı)"
        ]
        self.df['Zaman_Dilimi'] = pd.cut(self.df['Saat_Sayisal'], bins=bins, labels=labels, right=False)
        time_stats = self.df['Zaman_Dilimi'].value_counts().sort_index()
        return {"stats": time_stats.to_dict(), "most_active": time_stats.idxmax()}

    def get_daily_activity(self):
        """Gün analizi."""
        gun_cevirisi = {'Monday': 'Pazartesi', 'Tuesday': 'Salı', 'Wednesday': 'Çarşamba', 'Thursday': 'Perşembe', 'Friday': 'Cuma', 'Saturday': 'Cumartesi', 'Sunday': 'Pazar'}
        self.df['Gun'] = self.df['Tarih_Formatli'].dt.day_name().map(gun_cevirisi)
        daily_counts = self.df['Gun'].value_counts()
        return {"busiest_day": daily_counts.idxmax(), "max_count": int(daily_counts.max())}

    def get_deep_talks(self):
        """Kesintisiz sohbetler."""
        df_temp = self.df.copy()
        df_temp['Fark'] = df_temp['Tam_Zaman'].diff()
        threshold = pd.Timedelta(minutes=20)
        df_temp['Yeni_Oturum'] = df_temp['Fark'] > threshold
        df_temp['Oturum_ID'] = df_temp['Yeni_Oturum'].cumsum()
        sessions = df_temp.groupby('Oturum_ID').agg(Baslangic=('Tam_Zaman', 'min'), Bitis=('Tam_Zaman', 'max'), Mesaj_Sayisi=('Mesaj', 'count'))
        sessions['Sure'] = sessions['Bitis'] - sessions['Baslangic']
        deep_talks = sessions[(sessions['Sure'] > pd.Timedelta(hours=1)) & (sessions['Mesaj_Sayisi'] > 50)].sort_values('Sure', ascending=False)
        
        results = []
        for _, row in deep_talks.head(5).iterrows():
            total_seconds = row['Sure'].total_seconds()
            results.append({
                "date": row['Baslangic'].strftime('%d.%m.%Y'),
                "duration_str": f"{int(total_seconds // 3600)} sa {int((total_seconds % 3600) // 60)} dk",
                "message_count": int(row['Mesaj_Sayisi'])
            })
        return results

    def get_average_response_time(self):
        """Ortalama cevap süresi."""
        df_temp = self.df.copy()
        df_temp['Onceki_Gonderen'] = df_temp['Gonderen'].shift(1)
        df_temp['Onceki_Zaman'] = df_temp['Tam_Zaman'].shift(1)
        replies = df_temp[df_temp['Gonderen'] != df_temp['Onceki_Gonderen']].copy()
        replies['Cevap_Suresi'] = replies['Tam_Zaman'] - replies['Onceki_Zaman']
        valid_replies = replies[replies['Cevap_Suresi'] < pd.Timedelta(hours=12)]
        
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_replies = valid_replies[valid_replies['Gonderen'] == user]
            if not user_replies.empty:
                avg_seconds = user_replies['Cevap_Suresi'].mean().total_seconds()
                result[user] = {"minutes": int(avg_seconds // 60), "seconds": int(avg_seconds % 60)}
            else:
                result[user] = None
        return result

    def get_special_actions(self):
        """Silinen, düzenlenen mesajlar ve stickerlar."""
        targets = {
            "deleted": ["bu mesaj silindi", "bu ileti silindi"],
            "sticker": ["çıkartma dahil edilmedi", "sticker dahil edilmedi"],
            "edited": ["bu mesaj düzenlendi", "bu ileti düzenlendi", "<bu ileti düzenlendi>"]
        }
        result = {}
        for key, keywords in targets.items():
            pattern = "|".join(keywords)
            subset = self.df[self.df['Mesaj'].str.contains(pattern, case=False, na=False)]
            if not subset.empty:
                counts = subset['Gonderen'].value_counts()
                result[key] = {"leader": counts.idxmax(), "count": int(counts.max()), "all_stats": counts.to_dict()}
            else:
                result[key] = None
        return result

    def get_media_stats(self):
        """Sesli mesaj ve link analizi."""
        audio_patterns = ["sesli mesaj", "audio omitted", "ptt", "ses kaydı"]
        url_pattern = r'(https?://|www\.)\S+'
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str).str.lower()
            audio_count = user_msgs.apply(lambda x: any(p in x for p in audio_patterns)).sum()
            link_count = user_msgs.str.count(url_pattern).sum()
            result[user] = {"audio_count": int(audio_count), "link_count": int(link_count)}
        return result

    def get_laugh_style(self):
        """Gülüş stili."""
        patterns = {"random": r'(?:[asdfghjklşi]{4,})', "classic": r'(?:ha){3,}|(?:he){3,}|(?:eh){3,}', "short": r'\b(lol|kdkd|sjsj)\b'}
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str).str.lower()
            user_stats = {}
            for style, pattern in patterns.items():
                count = user_msgs.str.count(pattern).sum()
                if count > 0: user_stats[style] = int(count)
            result[user] = user_stats
        return result

    def get_initiator_stats(self):
        """Konu başlatıcı."""
        df_temp = self.df.copy()
        df_temp['Fark'] = df_temp['Tam_Zaman'].diff()
        initiations = df_temp[df_temp['Fark'] > pd.Timedelta(hours=6)]
        counts = initiations['Gonderen'].value_counts()
        if not counts.empty:
            return {"leader": counts.idxmax(), "stats": counts.to_dict()}
        return None

    def get_word_patterns(self):
        """İkili kelime kalıpları (Bigrams)."""
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str).str.lower()
            valid_bigrams = []
            for msg in user_msgs:
                if "bekleniyor" in msg and "alabilir" in msg: continue
                if "sildiniz" in msg: continue
                msg_clean = re.sub(r'[^\w\s]', '', msg)
                words = [w for w in msg_clean.split() if w not in STOP_WORDS and len(w) > 2]
                bigrams = list(zip(words, words[1:]))
                for bg in bigrams:
                    phrase = f"{bg[0]} {bg[1]}"
                    if any(bad in phrase for bad in BLACKLIST_BIGRAMS): continue
                    valid_bigrams.append(phrase)
            if valid_bigrams:
                result[user] = [{"phrase": p, "count": c} for p, c in Counter(valid_bigrams).most_common(10)]
            else:
                result[user] = []
        return result

    def get_question_stats(self):
        """Soru ve kelime uzunluğu."""
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str)
            question_count = user_msgs.str.count(r'\?').sum()
            avg_words = user_msgs.apply(lambda x: len(x.split())).mean()
            result[user] = {"question_count": int(question_count), "avg_word_length": round(avg_words, 1)}
        return result

    def get_peak_day(self):
        """Rekor gün."""
        daily_counts = self.df['Tarih'].value_counts()
        return {"date": daily_counts.idxmax(), "count": int(daily_counts.max())}

    def get_vocabulary_stats(self):
        """Kelime hazinesi."""
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            text = " ".join(self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str)).lower()
            text = re.sub(r'[^\w\s]', '', text)
            words = text.split()
            total = len(words)
            unique = len(set(words))
            ratio = (unique / total * 100) if total > 0 else 0
            result[user] = {"total": total, "unique": unique, "diversity_ratio": round(ratio, 2)}
        return result

    def get_milestones(self):
        """Yılın enleri: Geç saat, 1000. mesaj ve **En Kısa Mesaj**."""
        milestones = {}
        
        # En Geç Saat
        def shift_time(t_str):
            if pd.isna(t_str): return -1
            try:
                hour = int(t_str[:2])
                minute = int(t_str[3:5])
                if 0 <= hour < 6: hour += 24
                return hour * 60 + minute
            except: return -1
        self.df['Gece_Skoru'] = self.df['Saat'].apply(shift_time)
        latest_row = self.df.loc[self.df['Gece_Skoru'].idxmax()]
        milestones["latest_message"] = {"user": latest_row['Gonderen'], "time": latest_row['Saat'], "message": latest_row['Mesaj']}

        # 1000. Mesaj
        if len(self.df) >= 1000:
            msg_1000 = self.df.iloc[999]
            milestones["msg_1000"] = {"user": msg_1000['Gonderen'], "date": msg_1000['Tarih']}
        else:
            milestones["msg_1000"] = None

        # --- YENİ EKLENEN: En Kısa Mesaj ---
        # Sadece sistem mesajı olmayanları filtrele
        clean_df = self.df[~self.df['Mesaj'].str.lower().apply(lambda x: any(sf in x for sf in SYSTEM_FILTERS))]
        if not clean_df.empty:
            clean_df['Uzunluk'] = clean_df['Mesaj'].astype(str).apply(len)
            min_len = clean_df['Uzunluk'].min()
            shortest_row = clean_df[clean_df['Uzunluk'] == min_len].iloc[0]
            milestones["shortest_message"] = {"user": shortest_row['Gonderen'], "length": int(min_len), "message": shortest_row['Mesaj']}
        else:
            milestones["shortest_message"] = None
            
        return milestones

    def get_caps_lock_stats(self):
        """Caps lock analizi."""
        users = self.df['Gonderen'].unique()
        result = {}
        for user in users:
            user_msgs = self.df[self.df['Gonderen'] == user]['Mesaj'].astype(str)
            shout_count = 0
            for msg in user_msgs:
                clean_msg = re.sub(r'[^a-zA-ZğüşıöçĞÜŞİÖÇ]', '', msg)
                if len(clean_msg) > 3:
                    upper_count = sum(1 for c in clean_msg if c.isupper())
                    if upper_count / len(clean_msg) > 0.8: shout_count += 1
            result[user] = {"shout_count": shout_count}
        return result

File: src/utils/analyzer.js
import { STOP_WORDS, BLACKLIST_BIGRAMS, EMOJI_REGEX } from './constants';

export class WhatsAppAnalyzer {
  constructor(data) {
    this.df = data;
    this._preprocessData();
  }
  
  _preprocessData() {
    this.df = this.df.map(row => {
      const dateParts = row.Tarih.split(/[./-]/).map(Number);
      const d = dateParts[0];
      const m = dateParts[1] - 1; 
      let y = dateParts[2];
      if (y < 100) y += 2000;
      
      const timeParts = row.Saat.split(':').map(Number);
      const h = timeParts[0];
      const min = timeParts[1];
      const sec = timeParts[2] || 0;
      
      const tamZaman = new Date(y, m, d, h, min, sec);
      
      return {
        ...row,
        Tarih_Formatli: new Date(y, m, d),
        Tam_Zaman: tamZaman,
        Saat_Sayisal: h
      };
    }).filter(row => !isNaN(row.Tam_Zaman.getTime()))
    .filter(row => row.Tam_Zaman.getFullYear() === 2025);

    this.df.sort((a, b) => a.Tam_Zaman - b.Tam_Zaman);

    // Temiz Veri Seti (Sistem mesajları hariç)
    this.cleanDf = this.df.filter(row => !row.isSystem);
  }


  // --- ANALİZ FONKSİYONLARI ---

  getMessageStats() {
    const counts = {};
    this.df.forEach(row => {
      counts[row.Gonderen] = (counts[row.Gonderen] || 0) + 1;
    });
    return Object.entries(counts)
      .map(([kisi, sayi]) => ({ Kişi: kisi, 'Mesaj Sayısı': sayi }))
      .sort((a, b) => b['Mesaj Sayısı'] - a['Mesaj Sayısı']);
  }
  
  getTopWords(topN = 3) {
    const users = [...new Set(this.df.map(r => r.Gonderen))];
    const result = {};
    users.forEach(user => {
      const userMsgs = this.cleanDf.filter(r => r.Gonderen === user);
      const validWords = [];
      userMsgs.forEach(row => {
        const msgLower = row.Mesaj.toLowerCase();
        const msgClean = msgLower.replace(/[^\w\sğüşıöçĞÜŞİÖÇ]/g, '');
        const words = msgClean.split(/\s+/);
        words.forEach(w => {
          if (!STOP_WORDS.has(w) && w.length > 3) validWords.push(w);
        });
      });
      const counts = {};
      validWords.forEach(w => counts[w] = (counts[w] || 0) + 1);
      result[user] = Object.entries(counts).sort((a, b) => b[1] - a[1]).slice(0, topN).map(([word, count]) => ({ word, count }));
    });
    return result;
  }
  
  getEmojiStats(topN = 5) {
    const users = [...new Set(this.df.map(r => r.Gonderen))];
    const result = {};
    users.forEach(user => {
      const userMsgs = this.cleanDf.filter(r => r.Gonderen === user);
      const allEmojis = [];
      userMsgs.forEach(row => {
        const matches = row.Mesaj.match(EMOJI_REGEX);
        if (matches) allEmojis.push(...matches);
      });
      if (allEmojis.length > 0) {
        const counts = {};
        allEmojis.forEach(e => counts[e] = (counts[e] || 0) + 1);
        result[user] = Object.entries(counts).sort((a, b) => b[1] - a[1]).slice(0, topN).map(([emoji, count]) => ({ emoji, count }));
      } else {
        result[user] = [];
      }
    });
    return result;
  }
  
  getMediaStats() {
    const audioPatterns = ["sesli mesaj", "audio omitted", "ptt", "ses kaydı", "voice message"];
    const urlRegex = /(https?:\/\/|www\.)\S+/g;
    const users = [...new Set(this.df.map(r => r.Gonderen))];
    const result = {};
    users.forEach(user => {
        const userMsgs = this.df.filter(r => r.Gonderen === user);
        const audioCount = userMsgs.filter(row => audioPatterns.some(p => row.Mesaj.toLowerCase().includes(p))).length;
        const linkCount = userMsgs.reduce((acc, row) => {
            const matches = row.Mesaj.match(urlRegex);
            return acc + (matches ? matches.length : 0);
        }, 0);
        result[user] = { audio_count: audioCount, link_count: linkCount };
    });
    return result;
  }

  getHourlyActivity() {
    const bins = [[0, 3, "00-03 (Gece)"], [3, 6, "03-06 (Sabaha Karşı)"], [6, 9, "06-09 (Sabah)"], [9, 12, "09-12 (Kuşluk)"], [12, 15, "12-15 (Öğle)"], [15, 18, "15-18 (İkindi)"], [18, 21, "18-21 (Akşam)"], [21, 24, "21-00 (Gece Yarısı)"]];
    const stats = {};
    bins.forEach(([,,label]) => stats[label] = 0);
    this.df.forEach(row => {
        const h = row.Saat_Sayisal;
        for (let [start, end, label] of bins) {
            if (h >= start && h < end) { stats[label]++; break; }
        }
    });
    let mostActive = "-", maxVal = -1;
    Object.entries(stats).forEach(([k, v]) => { if (v > maxVal) { maxVal = v; mostActive = k; } });
    return { stats, most_active: mostActive };
  }
  
  getDailyActivity() {
    const gunCevirisi = ['Pazar', 'Pazartesi', 'Salı', 'Çarşamba', 'Perşembe', 'Cuma', 'Cumartesi'];
    const counts = {};
    this.df.forEach(row => {
      const gun = gunCevirisi[row.Tarih_Formatli.getDay()];
      counts[gun] = (counts[gun] || 0) + 1;
    });
    if (Object.keys(counts).length === 0) return { busiest_day: '-', max_count: 0 };
    const [busiestDay, maxCount] = Object.entries(counts).reduce((a, b) => a[1] > b[1] ? a : b);
    return { busiest_day: busiestDay, max_count: maxCount };
  }
  
  getDeepTalks() {
    if (this.df.length === 0) return [];
    const thresholdMs = 20 * 60 * 1000;
    const sessions = [];
    let currentSession = { start: this.df[0].Tam_Zaman, end: this.df[0].Tam_Zaman, msgCount: 1 };
    for (let i = 1; i < this.df.length; i++) {
        const diff = this.df[i].Tam_Zaman - this.df[i-1].Tam_Zaman;
        if (diff > thresholdMs) {
            sessions.push(currentSession);
            currentSession = { start: this.df[i].Tam_Zaman, end: this.df[i].Tam_Zaman, msgCount: 1 };
        } else {
            currentSession.end = this.df[i].Tam_Zaman;
            currentSession.msgCount++;
        }
    }
    sessions.push(currentSession);
    return sessions.filter(s => (s.end - s.start) > 3600000 && s.msgCount > 50)
        .sort((a, b) => (b.end - b.start) - (a.end - a.start)).slice(0, 5)
        .map(s => ({
            date: s.start.toLocaleDateString('tr-TR'),
            duration_str: `${Math.floor((s.end - s.start) / 3600000)} sa ${Math.floor(((s.end - s.start) % 3600000) / 60000)} dk`,
            message_count: s.msgCount
        }));
  }

  getAverageResponseTime() {
    const users = [...new Set(this.df.map(r => r.Gonderen))];
    const result = {};
    const replies = [];
    for (let i = 1; i < this.df.length; i++) {
      if (this.df[i].Gonderen !== this.df[i-1].Gonderen) {
        const diff = this.df[i].Tam_Zaman - this.df[i-1].Tam_Zaman;
        if (diff < 12 * 3600 * 1000) replies.push({ user: this.df[i].Gonderen, time: diff });
      }
    }
    users.forEach(user => {
        const userReplies = replies.filter(r => r.user === user);
        if (userReplies.length > 0) {
            const avgSec = (userReplies.reduce((acc, r) => acc + r.time, 0) / userReplies.length) / 1000;
            result[user] = { minutes: Math.floor(avgSec / 60), seconds: Math.floor(avgSec % 60) };
        } else { result[user] = null; }
    });
    return result;
  }
  
  getSpecialActions() {
      const targets = { deleted: ["bu mesaj silindi", "bu ileti silindi"], sticker: ["çıkartma dahil edilmedi", "sticker dahil edilmedi"], edited: ["bu mesaj düzenlendi", "bu ileti düzenlendi"] };
      const result = {};
      Object.entries(targets).forEach(([key, keywords]) => {
        const subset = this.df.filter(row => keywords.some(kw => row.Mesaj.toLowerCase().includes(kw)));
        
        const counts = {};
        if (subset.length > 0) {
          subset.forEach(row => counts[row.Gonderen] = (counts[row.Gonderen] || 0) + 1);
          const [leader, maxCount] = Object.entries(counts).reduce((a, b) => a[1] > b[1] ? a : b);
          // DÜZELTME: Veri olsa da olmasa da all_stats dönecek
          result[key] = { leader, count: maxCount, all_stats: counts };
        } else { 
            // DÜZELTME: Veri yoksa boş all_stats dön
            result[key] = { leader: '-', count: 0, all_stats: {} }; 
        }
      });
      return result;
  }

  getLaughStyle() {
      // DÜZELTME: Regex'ler genişletildi
      const patterns = { 
          random: /(?:[asdfghjklşi]{4,})/gi, 
          // 3 tekrardan 2 tekrara düşürdüm (haha, hehe kabul edilsin)
          classic: /(?:ha|he|eh|ehe|hak){2,}/gi, 
          short: /\b(lol|kdkd|sjsj|asds|ksks)\b/gi 
      };
      
      const users = [...new Set(this.df.map(r => r.Gonderen))];
      const result = {};
      users.forEach(user => {
          const userMsgs = this.cleanDf.filter(r => r.Gonderen === user);
          const styles = { random: 0, classic: 0, short: 0 };
          userMsgs.forEach(row => {
              const msg = row.Mesaj.toLowerCase();
              if (msg.match(patterns.random)) styles.random++;
              if (msg.match(patterns.classic)) styles.classic++;
              if (msg.match(patterns.short)) styles.short++;
          });
          result[user] = styles;
      });
      return result;
  }
  
  getInitiatorStats() {
      // DÜZELTME: Sohbet başlatıcı mantığı ve null kontrolü
      const threshold = 6 * 3600 * 1000; // 6 saat
      const counts = {};
      
      for (let i = 1; i < this.df.length; i++) {
          const diff = this.df[i].Tam_Zaman - this.df[i-1].Tam_Zaman;
          if (diff > threshold) {
              const initiator = this.df[i].Gonderen;
              counts[initiator] = (counts[initiator] || 0) + 1;
          }
      }
      
      if (Object.keys(counts).length > 0) {
          const [leader, count] = Object.entries(counts).reduce((a, b) => a[1] > b[1] ? a : b);
          return { leader, count, stats: counts };
      }
      return { leader: '-', count: 0, stats: {} };
  }
  
  getWordPatterns() {
      const users = [...new Set(this.df.map(r => r.Gonderen))];
      const result = {};
      users.forEach(user => {
          const userMsgs = this.cleanDf.filter(r => r.Gonderen === user);
          const validBigrams = [];
          userMsgs.forEach(row => {
              const msg = row.Mesaj.toLowerCase();
              if ((msg.includes("bekleniyor") && msg.includes("alabilir")) || msg.includes("sildiniz")) return;
              const words = msg.replace(/[^\w\sğüşıöçĞÜŞİÖÇ]/g, '').split(/\s+/).filter(w => !STOP_WORDS.has(w) && w.length > 2);
              for (let i = 0; i < words.length - 1; i++) {
                  const phrase = `${words[i]} ${words[i+1]}`;
                  if (!BLACKLIST_BIGRAMS.some(bad => phrase.includes(bad))) validBigrams.push(phrase);
              }
          });
          const counts = {};
          validBigrams.forEach(bg => counts[bg] = (counts[bg] || 0) + 1);
          result[user] = Object.entries(counts).sort((a, b) => b[1] - a[1]).slice(0, 10).map(([phrase, count]) => ({ phrase, count }));
      });
      return result;
  }
  
  getQuestionStats() {
      const users = [...new Set(this.df.map(r => r.Gonderen))];
      const result = {};
      users.forEach(user => {
          const userMsgs = this.cleanDf.filter(r => r.Gonderen === user);
          const avgWords = userMsgs.length > 0 ? (userMsgs.reduce((acc, row) => acc + row.Mesaj.split(/\s+/).length, 0) / userMsgs.length) : 0;
          result[user] = {
              question_count: userMsgs.reduce((acc, row) => acc + (row.Mesaj.includes('?') ? 1 : 0), 0),
              avg_word_length: parseFloat(avgWords.toFixed(1))
          };
      });
      return result;
  }
  
  getPeakDay() {
      const counts = {};
      this.df.forEach(row => counts[row.Tarih] = (counts[row.Tarih] || 0) + 1);
      const [bestDate, maxVal] = Object.entries(counts).reduce((a, b) => a[1] > b[1] ? a : b, ['-', 0]);
      return { date: bestDate, count: maxVal };
  }
  
  getVocabularyStats() {
      const users = [...new Set(this.df.map(r => r.Gonderen))];
      const result = {};
      users.forEach(user => {
          const words = this.cleanDf.filter(r => r.Gonderen === user).map(r => r.Mesaj).join(' ').toLowerCase().replace(/[^\w\sğüşıöçĞÜŞİÖÇ]/g, '').split(/\s+/).filter(w => w.length > 0);
          const total = words.length;
          result[user] = { total, unique: new Set(words).size, diversity_ratio: total > 0 ? parseFloat((new Set(words).size / total * 100).toFixed(2)) : 0 };
      });
      return result;
  }
  
  getMilestones() {
      let latestMsg = null, latestScore = -1;
      this.cleanDf.forEach(row => {
          let h = row.Saat_Sayisal;
          if (h >= 0 && h < 6) h += 24;
          const score = h * 60 + parseInt(row.Saat.split(':')[1]);
          if (score > latestScore) { latestScore = score; latestMsg = row; }
      });
      
      // DÜZELTME: 1000. mesaj ve tarih ekleme
      const msg1000Index = 999;
      const msg1000 = this.df.length > msg1000Index ? this.df[msg1000Index] : null;

      return {
          latest_message: latestMsg ? { 
              user: latestMsg.Gonderen, 
              time: latestMsg.Saat, 
              date: latestMsg.Tarih, // Tarih eklendi
              message: latestMsg.Mesaj 
          } : { user: '-', time: '-', date: '', message: '-' },
          msg_1000: msg1000 ? { user: msg1000.Gonderen, date: msg1000.Tarih } : null
      };
  }
  
  getCapsLockStats() {
      const users = [...new Set(this.df.map(r => r.Gonderen))];
      const result = {};
      users.forEach(user => {
          const userMsgs = this.cleanDf.filter(r => r.Gonderen === user);
          let shoutCount = 0;
          userMsgs.forEach(row => {
              const clean = row.Mesaj.replace(/[^a-zA-ZğüşıöçĞÜŞİÖÇ]/g, '');
              if (clean.length > 3 && (clean.split('').filter(c => c === c.toUpperCase()).length / clean.length) > 0.8) shoutCount++;
          });
          result[user] = { shout_count: shoutCount };
      });
      return result;
  }

  analyzeAll() {
    return {
      messages: this.getMessageStats(),
      words: this.getTopWords(),
      emojis: this.getEmojiStats(),
      hourly: this.getHourlyActivity(),
      daily: this.getDailyActivity(),
      deep_talks: this.getDeepTalks(),
      response_time: this.getAverageResponseTime(),
      special_actions: this.getSpecialActions(),
      media: this.getMediaStats(),
      laugh: this.getLaughStyle(),
      initiator: this.getInitiatorStats(), // Initiator eklendi
      bigrams: this.getWordPatterns(),
      questions: this.getQuestionStats(),
      peak_day: this.getPeakDay(),
      vocabulary: this.getVocabularyStats(),
      milestones: this.getMilestones(),
      caps: this.getCapsLockStats()
    };
  }
}